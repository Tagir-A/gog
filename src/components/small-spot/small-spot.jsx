import React, { Component } from "react";
import { PropTypes } from "prop-types";
import "./small-spot.css";
import { formatPrice } from "../../utils/formatters";

class SmallSpot extends Component {
  static propTypes = {
    game: PropTypes.shape({
      img: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      discount: PropTypes.number,
      owned: PropTypes.bool
    }).isRequired,
    isInCart: PropTypes.bool
  };

  render() {
    const { game, isInCart } = this.props;
    return (
      <div className="small-spot">
        <img src={game.img} alt={game.title} className="small-spot__image" />
        <div className="small-spot__info">
          <h3 className="small-spot__info__title">{game.title}</h3>
          <div className="small-spot__btns">
            {game.discount && (
              <div className="small-spot__discount">{`-${game.discount}%`}</div>
            )}
            {game.owned
              ? this.renderOwnedBtn()
              : isInCart
                ? this.renderInCartBtn()
                : this.renderAddBtn(game)}
          </div>
        </div>
      </div>
    );
  }
  renderAddBtn(game) {
    const {
      price: { value, minorUnits }
    } = game;
    const priceToShow = value / minorUnits;
    return (
      <button className="btn small-spot__btn" onClick={this.handleAddToCart}>
        {formatPrice(priceToShow)}
      </button>
    );
  }
  renderOwnedBtn() {
    return (
      <button
        className="btn small-spot__btn small-spot__btn--disabled"
        disabled
      >
        OWNED
      </button>
    );
  }

  renderInCartBtn() {
    return (
      <button className="btn small-spot__btn" disabled>
        IN CART
      </button>
    );
  }

  handleAddToCart = () => {
    const { onAddClick, game } = this.props;
    onAddClick && onAddClick(game);
  };
}

export { SmallSpot };
