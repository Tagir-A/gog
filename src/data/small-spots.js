import oddworld from '../assets/small-spots/oddworld.png'
import deponia from '../assets/small-spots/deponia.png'
import theSettlers2 from '../assets/small-spots/the-settlers-2.png'
import neverwinterNights from '../assets/small-spots/neverwinter-nights.png'
import assasinsCreed from '../assets/small-spots/assasins-creed.png'

const SMALL_SPOTS = [{
        title: "ODDWORLD: STRANGER'S WRATH",
        img: oddworld,
        price: {
            currency: 'USD',
            value: 999,
            minorUnits: 100
        },
        discount: 50
    },
    {
        title: "CHAOS ON DEPONIA",
        img: deponia,
        price: {
            currency: 'USD',
            value: 999,
            minorUnits: 100
        },
        owned: true
    },
    {
        title: "THE SETTLERS 2: GOLD EDITION",
        img: theSettlers2,
        price: {
            currency: 'USD',
            value: 599,
            minorUnits: 100
        }
    },
    {
        title: "NEVERWINTER NIGHTS",
        img: neverwinterNights,
        price: {
            currency: 'USD',
            value: 499,
            minorUnits: 100
        },
        discount: 50
    },
    {
        title: "ASSASIN'S CREED: DIRECTOR'S CUT",
        img: assasinsCreed,
        price: {
            currency: 'USD',
            value: 999,
            minorUnits: 100
        }
    }
]

export {
    SMALL_SPOTS
}
