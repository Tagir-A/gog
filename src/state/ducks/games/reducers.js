import {
    SMALL_SPOTS
} from "../../../data/small-spots";

const defaultState = [...SMALL_SPOTS]


export default function games(state = defaultState, action = {}) {
    return state;
}
