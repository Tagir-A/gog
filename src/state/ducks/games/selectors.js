export function isInCart(game, {
    products
}) {
    return Boolean(products.find(product => product.title === game.title));
}
