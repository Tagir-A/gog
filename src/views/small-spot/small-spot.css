.small-spot {
  display: flex;
  flex-direction: column;
  max-width: 196px;
  max-height: 185px;
  margin-right: 5px;
  margin-bottom: 20px;
  box-shadow: 0px 1px 5px 0 rgba(0, 0, 0, 0.15);
}

.small-spot:last-child {
  margin-right: 0;
}

.small-spot__image {
  display: block;
  max-height: 110px;
  width: 100%;
}

.small-spot__info {
  flex: 1 1 auto;
  background-color: #e0e0e0;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  padding: 11px 8px 8px 9px;
}

.small-spot__info__title {
  margin: 0;
  font-weight: var(--semi-bold);
  font-size: var(--main-font-size);
  color: var(--game-title-color);
}

.small-spot__btns {
  display: flex;
  justify-content: flex-end;
  align-items: center;
}

.small-spot__btn {
  font-size: var(--insignificant-font-size);
  font-weight: var(--semi-bold);
  color: var(--price-color);
  border: solid 1px rgba(0, 0, 0, 0.25);
  box-shadow: 0px 1px 3px 0 rgba(0, 0, 0, 0.25);
  min-width: 52px;
  height: 22px;
  opacity: var(--price-opacity);
}

.small-spot__btn--disabled {
  opacity: 0.25;
}

.small-spot__discount {
  width: 44px;
  height: 24px;
  margin-right: 6px;
  background-color: var(--discount-color);
  color: #f2f2f2;
  display: flex;
  justify-content: center;
  align-items: center;
  font-size: var(--insignificant-font-size);
  font-weight: var(--semi-bold);
  border-radius: var(--button-border-radius);
}
