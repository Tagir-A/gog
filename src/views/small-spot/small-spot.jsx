import React, { Component } from "react";
import PropTypes from "prop-types";
import "./small-spot.css";
import { formatPrice } from "../../utils/formatters";
import game from "../../models/game";

function SmallSpot({ game, onAddClick }) {
  const handleAddToCart = () => {
    onAddClick && onAddClick(game);
  };
  function renderAddBtn(game) {
    const {
      price: { value, minorUnits }
    } = game;
    const priceToShow = value / minorUnits;
    return (
      <button className="btn small-spot__btn" onClick={handleAddToCart}>
        {formatPrice(priceToShow)}
      </button>
    );
  }
  function renderOwnedBtn() {
    return (
      <button
        className="btn small-spot__btn small-spot__btn--disabled"
        disabled>
        OWNED
      </button>
    );
  }

  function renderInCartBtn() {
    return (
      <button className="btn small-spot__btn" disabled>
        IN CART
      </button>
    );
  }

  return (
    <div className="small-spot">
      <img src={game.img} alt={game.title} className="small-spot__image" />
      <div className="small-spot__info">
        <h3 className="small-spot__info__title">{game.title}</h3>
        <div className="small-spot__btns">
          {game.discount && (
            <div className="small-spot__discount">{`-${game.discount}%`}</div>
          )}
          {game.owned
            ? renderOwnedBtn()
            : game.isInCart
              ? renderInCartBtn()
              : renderAddBtn(game)}
        </div>
      </div>
    </div>
  );
}

SmallSpot.propTypes = {
  game: game.isRequired,
  onAddClick: PropTypes.func.isRequired
};

export { SmallSpot };
