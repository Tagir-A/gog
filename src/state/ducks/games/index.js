import reducer from "./reducers";
import * as gamesSelectors from "./selectors";

export {
    gamesSelectors
}
export default reducer;
