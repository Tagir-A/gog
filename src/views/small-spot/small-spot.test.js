import { SmallSpot } from "./small-spot";
import { GAMES } from "../../data/games";

const props = {
  game: GAMES[0],
  onAddClick: () => null
};

it("renders without crashing", () => {
  const wrapper = shallow(<SmallSpot {...props} />);
  expect(wrapper.debug()).toMatchSnapshot();
});

it("renders owned button when game is owned", () => {
  const innerProps = {
    ...props,
    game: {
      ...GAMES[0],
      owned: true
    }
  };
  const wrapper = shallow(<SmallSpot {...innerProps} />);
  expect(wrapper.debug()).toMatchSnapshot();
  expect(wrapper.find(".small-spot__btn--disabled").text()).toEqual("OWNED");
});

it("renders add button when game not owned and not in a cart", () => {
  const wrapper = shallow(<SmallSpot {...props} />);
  expect(wrapper.debug()).toMatchSnapshot();
});
