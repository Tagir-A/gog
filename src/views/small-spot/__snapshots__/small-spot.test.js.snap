// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders add button when game not owned and not in a cart 1`] = `
"<div className=\\"small-spot\\">
  <img src=\\"oddworld.png\\" alt=\\"ODDWORLD: STRANGER\\\\'S WRATH\\" className=\\"small-spot__image\\" />
  <div className=\\"small-spot__info\\">
    <h3 className=\\"small-spot__info__title\\">
      ODDWORLD: STRANGER&#39;S WRATH
    </h3>
    <div className=\\"small-spot__btns\\">
      <div className=\\"small-spot__discount\\">
        -50%
      </div>
      <button className=\\"btn small-spot__btn\\" onClick={[Function: handleAddToCart]}>
        $9.99
      </button>
    </div>
  </div>
</div>"
`;

exports[`renders owned button when game is owned 1`] = `
"<div className=\\"small-spot\\">
  <img src=\\"oddworld.png\\" alt=\\"ODDWORLD: STRANGER\\\\'S WRATH\\" className=\\"small-spot__image\\" />
  <div className=\\"small-spot__info\\">
    <h3 className=\\"small-spot__info__title\\">
      ODDWORLD: STRANGER&#39;S WRATH
    </h3>
    <div className=\\"small-spot__btns\\">
      <div className=\\"small-spot__discount\\">
        -50%
      </div>
      <button className=\\"btn small-spot__btn small-spot__btn--disabled\\" disabled={true}>
        OWNED
      </button>
    </div>
  </div>
</div>"
`;

exports[`renders without crashing 1`] = `
"<div className=\\"small-spot\\">
  <img src=\\"oddworld.png\\" alt=\\"ODDWORLD: STRANGER\\\\'S WRATH\\" className=\\"small-spot__image\\" />
  <div className=\\"small-spot__info\\">
    <h3 className=\\"small-spot__info__title\\">
      ODDWORLD: STRANGER&#39;S WRATH
    </h3>
    <div className=\\"small-spot__btns\\">
      <div className=\\"small-spot__discount\\">
        -50%
      </div>
      <button className=\\"btn small-spot__btn\\" onClick={[Function: handleAddToCart]}>
        $9.99
      </button>
    </div>
  </div>
</div>"
`;
